From 973aa6abc8a6fb479e8cc6f8161568586e7cfe74 Mon Sep 17 00:00:00 2001
From: Muhamad Sammar <muhamad.sammar@arm.com>
Date: Thu, 3 Sep 2026 09:04:38 +0000
Subject: [PATCH] Preserve old swapchain image indices

Release unused image resources in place when retiring a swapchain so indices held by applications remain valid. Treat released slots as non-pending during teardown.

Change-Id: Ib90a7c2ac2033b1c863b06dd00be691732665a44

Upstream-Status: Backport [https://gitlab.freedesktop.org/mesa/vulkan-wsi-layer/-/commit/973aa6abc8a6fb479e8cc6f8161568586e7cfe74]
---
 wsi/swapchain_base.cpp  | 25 ++++++++++++++++++-------
 wsi/swapchain_image.hpp |  8 +++++---
 2 files changed, 23 insertions(+), 10 deletions(-)

diff --git a/wsi/swapchain_base.cpp b/wsi/swapchain_base.cpp
index eb5f0f1..c78ac5b 100644
--- a/wsi/swapchain_base.cpp
+++ b/wsi/swapchain_base.cpp
@@ -37,7 +37,6 @@
 #include <cstdio>
 #include <cstdlib>
 #include <system_error>
-#include <algorithm>
 
 #include <unistd.h>
 #include <vulkan/vulkan.h>
@@ -696,10 +695,21 @@ VkResult swapchain_base::queue_present(VkQueue queue, const VkPresentInfoKHR *pr
 
 void swapchain_base::deprecate(VkSwapchainKHR descendant)
 {
-   /* Remove any images that have not been acquried or presented */
-   m_swapchain_images.erase(std::remove_if(m_swapchain_images.begin(), m_swapchain_images.end(),
-                                           [](auto &image) { return image.get_status() == swapchain_image::FREE; }),
-                            m_swapchain_images.end());
+   util::unique_lock<util::recursive_mutex> image_status_lock(m_image_status_mutex);
+   if (!image_status_lock)
+   {
+      WSI_LOG_ERROR("Failed to acquire mutex lock in deprecate.\n");
+      abort();
+   }
+
+   /* Release unused resources without changing Vulkan-visible image indexes. */
+   for (auto &image : m_swapchain_images)
+   {
+      if (image.get_status() == swapchain_image::FREE)
+      {
+         image.destroy();
+      }
+   }
 
    /* Set its descendant. */
    m_descendant = descendant;
@@ -724,9 +734,10 @@ void swapchain_base::wait_for_pending_buffers()
        * and then subsequent calls might block if the semaphore value
        * has reached to zero.
        */
-      if ((img.get_status() == swapchain_image::ACQUIRED) || (img.get_status() == swapchain_image::FREE))
+      if ((img.get_status() == swapchain_image::ACQUIRED) || (img.get_status() == swapchain_image::FREE) ||
+          (img.get_status() == swapchain_image::UNALLOCATED))
       {
-         /* If the image is acquired or free, it is not pending. */
+         /* If the image is acquired, free, or unallocated, it is not pending. */
          non_pending_images++;
       }
    }
diff --git a/wsi/swapchain_image.hpp b/wsi/swapchain_image.hpp
index 9602a6b..3ca04c6 100644
--- a/wsi/swapchain_image.hpp
+++ b/wsi/swapchain_image.hpp
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2025 Arm Limited.
+ * Copyright (c) 2025-2026 Arm Limited.
  *
  * SPDX-License-Identifier: MIT
  *
@@ -47,6 +47,7 @@ namespace wsi
 {
 
 class swapchain_image_factory;
+class swapchain_base;
 
 /**
  * @brief Base class describing swapchain image memory allocator/binder
@@ -256,8 +257,9 @@ public:
    }
 
 private:
-   /* Only allow swapchain image factory to create this class and acquire backing memory from it */
+   /* Only allow swapchain internals to create or destroy image resources. */
    friend swapchain_image_factory;
+   friend swapchain_base;
 
    /**
     * @brief Construct a new swapchain image object
@@ -378,4 +380,4 @@ public:
    virtual util::unique_ptr<image_backing_memory> create_image_backing_memory(util::allocator &allocator) = 0;
 };
 
-}
\ No newline at end of file
+}
